## Exercise 02 - option price introduction

The purpose of this is to introduce you to a small amount of options data.  You'll perform a few calculations (which aren't all that meaningful) but the real motivation is to get your hand a little dirty with options data.

Along the way, you will use some of the *indexing* and *masking* skills that you learned in Tutorials 04 and 05.

We will use the data in following location: *data/spy_put.csv*

#### 1) Import the `pandas` and `numpy` packages.

In [1]:
import pandas as pd
import numpy as np

#### 2) Read-in the data into a variable call `df_put`.

In [3]:
df_put = pd.read_csv('data/spy_put.csv')
df_put

underlying_symbol  underlying_price type  expiration   data_date  d2x  \
0                SPY        276.179688  put  2018-11-16  2018-10-19   20   
1                SPY        275.029999  put  2018-11-16  2018-10-22   19   
2                SPY        273.519989  put  2018-11-16  2018-10-23   18   
3                SPY        265.209991  put  2018-11-16  2018-10-24   17   
4                SPY        270.010010  put  2018-11-16  2018-10-25   16   
5                SPY        265.220001  put  2018-11-16  2018-10-26   15   
6                SPY        263.690002  put  2018-11-16  2018-10-29   14   
7                SPY        267.880005  put  2018-11-16  2018-10-30   13   
8                SPY        270.690002  put  2018-11-16  2018-10-31   12   
9                SPY        273.369995  put  2018-11-16  2018-11-01   11   
10               SPY        271.799988  put  2018-11-16  2018-11-02   10   
11               SPY        273.470001  put  2018-11-16  2018-11-05    9   
12               SPY        275.109985  put  2018-11-16  2018-11-06    8   
13               SPY        281.000000  put  2018-11-16  2018-11-07    7   
14               SPY        280.480011  put  2018-11-16  2018-11-08    6   
15               SPY        277.820007  put  2018-11-16  2018-11-09    5   
16               SPY        272.479492  put  2018-11-16  2018-11-12    4   
17               SPY        272.040009  put  2018-11-16  2018-11-13    3   
18               SPY        270.149994  put  2018-11-16  2018-11-14    2   
19               SPY        273.010010  put  2018-11-16  2018-11-15    1   
20               SPY        273.730011  put  2018-11-16  2018-11-16    0   

    strike        bid        ask        mid  volume  open_interest  \
0      274   4.240000   4.270000   4.255000    7027          58931   
1      274   4.280000   4.310000   4.295000    3174          58861   
2      274   5.090000   5.140000   5.115000   32932          59467   
3      274   9.590000  10.200000   9.895000   11461          49153   
4      274   8.110000   8.190000   8.150000    3394          44592   
5      274   9.930000  10.260000  10.095000    2387          44294   
6      274  11.390000  11.860000  11.625000     741          43408   
7      274   8.290000   8.370000   8.330000    1365          43202   
8      274   5.940000   6.000000   5.970000    1392          42929   
9      274   4.240000   4.270000   4.255000    2677          42731   
10     274   4.760000   4.830000   4.795000    3379          42250   
11     274   3.660000   3.690000   3.675000    1463          42069   
12     274   2.830000   2.870000   2.850000    7659          41989   
13     274   0.690000   0.700000   0.695000   14659          42503   
14     274   0.660000   0.670000   0.665000    4329          44655   
15     274   1.070000   1.090000   1.080000   19481          45384   
16     274   2.960000   3.010000   2.985000   19070          45631   
17     274   3.130000   3.200000   3.165000   11396          43884   
18     274   4.630000   4.740000   4.685000   21398          45396   
19     274   1.750000   1.810000   1.780000   12187          43941   
20     274   0.269989   0.269989   0.269989   45242          40189   

    implied_vol     delta  
0      0.174777  0.417679  
1      0.167284  0.441894  
2      0.176492  0.488031  
3      0.190524  0.703797  
4      0.196509  0.632242  
5      0.202849  0.711129  
6      0.193238  0.792172  
7      0.199422  0.673333  
8      0.185986  0.592990  
9      0.175946  0.511377  
10     0.178116  0.564026  
11     0.177214  0.494581  
12     0.185068  0.421302  
13     0.164618  0.167157  
14     0.166948  0.170041  
15     0.162767  0.272960  
16     0.174252  0.568382  
17     0.173829  0.640909  
18     0.194683  0.811336  
19     0.179841  0.620476  
20     0.000000  1.000000

For those of you unfamiliar with options, this might seem a little overwhelming, but don't worry.  You will be able to complete the exercises without any prior knowledge of options.

Options are derivatives contracts written on top of shares of stocks.  Unlike stocks, options have an expiration date - on that date they either have some positive payout, or a they are worth nothing.  After expiration they cease to exist.  We are looking at end-of-day prices, for one particular option, for the 21 days before it expires.

In particular, this is **put** option that has SPY as its **underlying**. The **expiration date** of our option is 11/16/2018.  The **strike price** of the contract is 274.  

Don't worry if this all sounds like gibberish.

#### 3) Use fancy indexing to create a `DataFrame` that consists of the put's contract features: `underlying_symbol`, `type`, `expiration`, and `strike`.  This won't be a very interesting `DataFrame` because all 21 rows will be the same.

In [9]:
df_put[['underlying_symbol', 'type', 'expiration', 'strike']] # alternative 1
#df_put.loc[:, ['underlying_symbol', 'type', 'expiration', 'strike']] # alternative 2

underlying_symbol type  expiration  strike
0                SPY  put  2018-11-16     274
1                SPY  put  2018-11-16     274
2                SPY  put  2018-11-16     274
3                SPY  put  2018-11-16     274
4                SPY  put  2018-11-16     274
5                SPY  put  2018-11-16     274
6                SPY  put  2018-11-16     274
7                SPY  put  2018-11-16     274
8                SPY  put  2018-11-16     274
9                SPY  put  2018-11-16     274
10               SPY  put  2018-11-16     274
11               SPY  put  2018-11-16     274
12               SPY  put  2018-11-16     274
13               SPY  put  2018-11-16     274
14               SPY  put  2018-11-16     274
15               SPY  put  2018-11-16     274
16               SPY  put  2018-11-16     274
17               SPY  put  2018-11-16     274
18               SPY  put  2018-11-16     274
19               SPY  put  2018-11-16     274
20               SPY  put  2018-11-16     274

#### 4) Create a `Series` of booleans, call it `ser_itm`,  by performing a pair-wise comparison of  `df_put['underlying_price']` and `df_put['strike']`.  In particular, check when the underlying price is less than (`<`) the strike price of the option. 

In [18]:
ser_itm = df_put['underlying_price'] < df_put['strike']
ser_itm

0     False
1     False
2      True
3      True
4      True
5      True
6      True
7      True
8      True
9      True
10     True
11     True
12    False
13    False
14    False
15    False
16     True
17     True
18     True
19     True
20     True
dtype: bool

#### 5) A put is said to be "in-the-money" if the price of the underlying is less than strike of the put.  Use masking to select only the rows from `df_put` when the contract is in the money.

In [19]:
df_put[ser_itm]

underlying_symbol  underlying_price type  expiration   data_date  d2x  \
2                SPY        273.519989  put  2018-11-16  2018-10-23   18   
3                SPY        265.209991  put  2018-11-16  2018-10-24   17   
4                SPY        270.010010  put  2018-11-16  2018-10-25   16   
5                SPY        265.220001  put  2018-11-16  2018-10-26   15   
6                SPY        263.690002  put  2018-11-16  2018-10-29   14   
7                SPY        267.880005  put  2018-11-16  2018-10-30   13   
8                SPY        270.690002  put  2018-11-16  2018-10-31   12   
9                SPY        273.369995  put  2018-11-16  2018-11-01   11   
10               SPY        271.799988  put  2018-11-16  2018-11-02   10   
11               SPY        273.470001  put  2018-11-16  2018-11-05    9   
16               SPY        272.479492  put  2018-11-16  2018-11-12    4   
17               SPY        272.040009  put  2018-11-16  2018-11-13    3   
18               SPY        270.149994  put  2018-11-16  2018-11-14    2   
19               SPY        273.010010  put  2018-11-16  2018-11-15    1   
20               SPY        273.730011  put  2018-11-16  2018-11-16    0   

    strike        bid        ask        mid  volume  open_interest  \
2      274   5.090000   5.140000   5.115000   32932          59467   
3      274   9.590000  10.200000   9.895000   11461          49153   
4      274   8.110000   8.190000   8.150000    3394          44592   
5      274   9.930000  10.260000  10.095000    2387          44294   
6      274  11.390000  11.860000  11.625000     741          43408   
7      274   8.290000   8.370000   8.330000    1365          43202   
8      274   5.940000   6.000000   5.970000    1392          42929   
9      274   4.240000   4.270000   4.255000    2677          42731   
10     274   4.760000   4.830000   4.795000    3379          42250   
11     274   3.660000   3.690000   3.675000    1463          42069   
16     274   2.960000   3.010000   2.985000   19070          45631   
17     274   3.130000   3.200000   3.165000   11396          43884   
18     274   4.630000   4.740000   4.685000   21398          45396   
19     274   1.750000   1.810000   1.780000   12187          43941   
20     274   0.269989   0.269989   0.269989   45242          40189   

    implied_vol     delta  
2      0.176492  0.488031  
3      0.190524  0.703797  
4      0.196509  0.632242  
5      0.202849  0.711129  
6      0.193238  0.792172  
7      0.199422  0.673333  
8      0.185986  0.592990  
9      0.175946  0.511377  
10     0.178116  0.564026  
11     0.177214  0.494581  
16     0.174252  0.568382  
17     0.173829  0.640909  
18     0.194683  0.811336  
19     0.179841  0.620476  
20     0.000000  1.000000

#### 6) Calculate the average delta for days that the put is in-the-money, and also for the days that it is out-of-the-money.  (Hint: find out how to "negate" an array of booleans, apply your masking skills, and then apply your aggreation function skills) 

In [13]:
df_put[ser_itm]['delta'].mean()

0.653652024181308

In [20]:
df_put[~(ser_itm)]['delta'].mean()

0.31517216213921667